Solaris 10 Fixes

gsauthof opened the pull request:
Fixes #1193 and #1194, a double free, some undefined behavior and some warnings.

gsauthof committed:
Fix nice() for realtime processes under Solaris 10

fixes #1194

gsauthof committed:
Use psinfo as fallback [gu]id source on Solaris 10

fixes #1193

gsauthof committed:
Fix double free

gsauthof committed:
Match ssize_t return type of read functions

gsauthof committed:
Fix undefined behavior

with respect to strict aliasing rules and fix some warnings

For example, under strict aliasing rules of the C standard, casting a
char pointer to a struct pointer and accessing the character array
through that struct pointer yields undefined behavior.

giampaolo commented on psutil/_pssunos.py:
```diff
@@ -432,27 +439,10 @@ def num_threads(self):
+        # Note #1: getpriority(3) doesn't work for realtime processes.
+        # Psinfo is what ps uses, see:
+        # https://github.com/giampaolo/psutil/issues/1194
+        return self._proc_basic_info()[proc_info_map['nice']]
```
Nice that we got rid of that mess.

giampaolo commented on psutil/_psutil_common.c:
```diff
@@ -40,7 +40,7 @@ PyUnicode_DecodeFSDefaultAndSize(char *s, Py_ssize_t size) {
  */
 PyObject *
-NoSuchProcess(char *msg) {
+NoSuchProcess(const char *msg) {
```
why is this necessary?

gsauthof replied:
`const char *` means pointer to const characters. The const modifier signals the caller that NoSuchProcess will not change the passed message. Otherwise, it's not clear and one has to inspect the function body. In case someone accidentally writes to msg using that const pointer it's a compile error. This change also remove compiler warnings since NoSuchProcess is called with string literals that effectively have the type `const char *`. Thus, this additional type safety protects against accidental writing into a const character array through a non-const pointer which would invoke undefined behavior (e.g. crash if the string literals are placed in a read-only data section).

giampaolo commented on psutil/_psutil_common.h:
```diff
@@ -4,6 +4,9 @@
  */
 
+#ifndef PSUTIL_PSUTIL_COMMON_H
+#define PSUTIL_PSUTIL_COMMON_H
```
why is this necessary?

gsauthof replied:
This guards the header from being included multiple times into one translation unit (either directly/indirectly). In C, unless the header does something special, you always define such guards (or use the non-standardized once pragma).

giampaolo commented on psutil/_pssunos.py:
```diff
@@ -394,7 +398,10 @@ def _proc_basic_info(self):
     @memoize_when_activated
     def _proc_cred(self):
-        return cext.proc_cred(self.pid, self._procfs_path)
+        @wrap_exceptions
```
Why not decorate the method instead?

gsauthof replied:
I tried stacking the `@wrap_exceptions` on top of `@memoize_when_activated` of `_proc_cred` but it didn't work. Something about a missing attribute, IIRC.

giampaolo commented on psutil/_pssunos.py:
```diff
@@ -471,12 +461,22 @@ def ppid(self):
+        except AccessDenied:
+            real      = self._proc_basic_info()[proc_info_map['gid']]
+            effective = self._proc_basic_info()[proc_info_map['egid']]
+            saved     = None
```
PEP8 (please remove extra spaces)

gsauthof replied:
Ok, will update (rebase) the commit.

giampaolo commented on psutil/_pssunos.py:
```diff
@@ -471,12 +461,22 @@ def ppid(self):
+        except AccessDenied:
+            real      = self._proc_basic_info()[proc_info_map['uid']]
+            effective = self._proc_basic_info()[proc_info_map['euid']]
+            saved     = None
```
PEP8 (please remove extra spaces)

gsauthof replied:
Ok, will fix this.

giampaolo reviewed (changes requested):
Nice work. Could you please update HISTORY file describing the fixes/improvements which occurred in this PR?

gsauthof committed:
Update HISTORY with Solaris notes

gsauthof commented:
So I've updated the pull request and implemented the requested changes (HISTORY and PEP 8).

Btw, looking at the `HISTORY.rst` file I noticed:

    - 1120_: [SunOS] cmdline() could be truncated at the 15th character when
      reading it from /proc. An extra effort is made by reading it from process
      address space first.

But #1120 doesn't really match the description. Probably a mixup with #694, right?

giampaolo commented:
> But #1120 doesn't really match the description. Probably a mixup with #694, right?

correct; thanks for the PR.
